openshift/catalog-sync/pkg/extract/extract.go

```diff
@@ -0,0 +1,180 @@
+func pullImageToLayout(ctx context.Context, imageRef, tmpDir string) error {
+	srcRef, err := docker.ParseReference("//" + imageRef)
+	if err != nil {
+		return fmt.Errorf("parse image ref: %w", err)
+	}
+
+	destRef, err := layout.ParseReference(fmt.Sprintf("%s:%s", tmpDir, "v1"))
+	if err != nil {
+		return fmt.Errorf("parse layout ref: %w", err)
+	}
+
+	// This code is used for **dev/test only**
+	// On Linux systems with tools like podman, buildah, or skopeo, a default policy.json may
+	// be available at ~/.config/containers or /etc/containers.
+	// However, this is not the case on mac/docker envs.
+	// To avoid unnecessary setup complexity, we explicitly skip policy validation here.
+	policy := &signature.Policy{
+		Default: []signature.PolicyRequirement{
+			signature.NewPRInsecureAcceptAnything(),
+		},
+	}
+	policyCtx, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+	defer policyCtx.Destroy()
+
+	// Force image resolution to Linux platform to avoid OS mismatch errors
+	// during testing on Mac. Example error:
+	// err: <*errors.errorString>{
+	//     s: "no image found in manifest list for architecture \"arm64\", variant \"v8\", OS \"darwin\"",
+	// }
+	sysCtx := &types.SystemContext{
+		OSChoice: "linux",
+	}
+
+	if _, err := imgcopy.Image(ctx, policyCtx, destRef, srcRef, &imgcopy.Options{
+		SourceCtx:      sysCtx,
+		DestinationCtx: sysCtx,
+	}); err != nil {
+		return fmt.Errorf("copy image: %w", err)
+	}
+
+	return nil
+}
+
+// extractFilesystem extracts the filesystem layers from the OCI image into <tmpDir>/fs
+func extractFilesystem(ctx context.Context, ociDir string) error {
+	ref, err := layout.ParseReference(fmt.Sprintf("%s:%s", ociDir, "v1"))
+	if err != nil {
+		return fmt.Errorf("parse layout: %w", err)
+	}
+	src, err := ref.NewImageSource(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("open image source: %w", err)
+	}
+	defer src.Close()
+
+	manifests, _, err := src.GetManifest(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("get manifest: %w", err)
+	}
+
+	mf, err := manifest.FromBlob(manifests, manifest.GuessMIMEType(manifests))
+	if err != nil {
+		return fmt.Errorf("parse manifest: %w", err)
+	}
+
+	dir := filepath.Join(ociDir, "fs")
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("make fs dir: %w", err)
+	}
+
+	for i, layer := range mf.LayerInfos() {
+		rc, _, err := src.GetBlob(ctx, layer.BlobInfo, nil)
+		if err != nil {
+			return fmt.Errorf("get blob %d: %w", i, err)
+		}
+
+		decompressed, _, err := compression.AutoDecompress(rc)
+		if err != nil {
+			rc.Close()
+			return fmt.Errorf("decompress blob %d: %w", i, err)
+		}
+
+		// To avoid permission errors faced
+		mask := os.FileMode(0755)
+		opts := &archive.TarOptions{
+			NoLchown:          true,
+			IgnoreChownErrors: true,
+			InUserNS:          true,
+			ForceMask:         &mask,
+		}
```

Comment:
It would be helpful for each of these options to have comments explaining their purpose.

Reply:
I will add them 👍

Reply:
I added them one by one.
But the latest one seems to solve all needs.
So, I could clean up it all.